### PR TITLE
002 normalise graph

### PR DESCRIPTION
**Pull request by aknipey**
Added normalise and percent change sliders, with corresponding axis name changes
Screenshot from 2023-09-07 19-17-44

### CODE REVIEW

File: frontend/src/components/OutcomeGraph.tsx [hunk 1]
```diff
@@ -91,48 +105,111 @@ export function OutcomeGraph(props: {
...
+          if (normalised && percentageChangeAxis) {
+            return portfolioState.realised / outcome[0].realised * 100 - 100;
+          } else if (normalised && !percentageChangeAxis) {
+            return portfolioState.realised - outcome[0].realised;
+          } else if (!normalised && percentageChangeAxis) {
+            return portfolioState.realised / outcome[0].realised * 100;
+          } else {
+            return portfolioState.realised;
+          }
+        }),
```
Comment (patrickemoore) [inline frontend/src/components/OutcomeGraph.tsx:118]: This is very hacky. I think we can represent this as linear transformation. So we should apply the normalisation if it is normalised, then apply the scaling. Whatever it is, this case bash should be simplified.

Comment (patrickemoore) [reply frontend/src/components/OutcomeGraph.tsx:118]: Also extract helper functions, these might be useful later

File: frontend/src/components/OutcomeGraph.tsx [hunk 2]
```diff
@@ -91,48 +105,111 @@ export function OutcomeGraph(props: {
...
+    },
+    yaxis: y,
+  })}, [normalised, percentageChangeAxis]);
 
   const markerPlotlyGraphs = useMemo(() => {
     return markerGraphs?.map(outcome => {
       return {
         x: outcome.map((_, i) => i),
-        y: outcome.map(portfolioState => portfolioState.realised),
+        y: outcome.map(portfolioState => {
```
Comment (patrickemoore) [inline frontend/src/components/OutcomeGraph.tsx:186]: As above

File: frontend/src/components/OutcomeGraph.tsx [hunk 3]
```diff
@@ -189,7 +266,17 @@ export function OutcomeGraph(props: {
...
+              <Stack direction='column' alignItems='center'>
+                <Typography variant='body1'>Normalised:</Typography>
+                <Switch checked={normalised} onChange={toggleNormalised} />
+              </Stack>
+              <Box width={30} />
+              <Stack direction='column' alignItems='center'>
+                <Typography variant='body1'>Percent Change:</Typography>
+                <Switch checked={percentageChangeAxis} onChange={togglePercentageChangeAxis} />
+              </Stack>
+            </Stack>
```
Comment (patrickemoore) [inline frontend/src/components/OutcomeGraph.tsx:279]: This indentation is off, does eslint not pick this up?

File: frontend/src/components/OutcomeGraph.tsx (hunk 1 above)
Comment (patrickemoore) [reply frontend/src/components/OutcomeGraph.tsx:118]: (btw defining helper functions as constants in the global scope is much more efficient in react as react doesn't have to recalculate the function every time)

**Review by patrickemoore — CHANGES_REQUESTED** [review]
almost lgtm, but need to streamline the normalisation and scaling math. Performance is a reasonably big concern for the 1000 simuations.

File: frontend/src/components/OutcomeGraph.tsx (hunk 1 above)
Comment (aknipey) [reply frontend/src/components/OutcomeGraph.tsx:118]: Oh ok yeah ill do that. Ill put a function in utils
Are there inbuilt tools for linear transformations or do i need a library

Comment (patrickemoore) [reply frontend/src/components/OutcomeGraph.tsx:118]: just do the maths yourself

if (normalised) normalise();
if (scaled) scale();

Comment (aknipey) [reply frontend/src/components/OutcomeGraph.tsx:118]: yep thats what i started doing ok cool

Comment (aknipey) [reply frontend/src/components/OutcomeGraph.tsx:118]: oh this is actually so much easier

**Review by patrickemoore — APPROVED** [review]

**Pull request merged**